src/pool/cli.c

```diff
@@ -1507,6 +1507,12 @@ create_map_refresh_rpc(struct dc_pool *pool, unsigned int map_version,
 		return rc;
 	}
 
+	/* Shorten the timeout to speed up map_refresh tasks. */
+	rc = crt_req_get_timeout(c, &timeout);
```

Comment:
probably get_timeout should use ctxt as the parameter?

Reply:
@wangdi1, could you elaborate a bit on this, please? (Notice that the intention is to get the timeout value of _this request_, not of the context or of the global configuration.)

Reply:
this RPC is just created, so I assume it will use default cart timeout from the ctx?  so your intention is to keep shrinking the timeout if RPC is timeout? or just shrinking the general timeout of the pool map refresh.

Hmm, another question would be who should manage the timeout? DAOS or cart? per module or context?  we probably need work out some protocol here? sorry, this is definitely not this patch issue.  But once you provide such api, someone else might use it wildly.


Btw: could you please elaborate the issue you saw?

Reply:
@wangdi1, the intention is to just shrink the general timeout of POOL_TGT_QUERY_MAP RPCs. If one such RPC times out, next one is meant to use the same timeout. crt_req_get_timeout enables crt users to know the timeout of an RPC without worrying about if a context-wise timeout has been configured or how a context-wise timeout relates to the global timeout.

As to who should manage timeouts, I assume you mean the "set of principles for RPC and event timeouts" I mentioned below in the comment about "1/4". My understanding is probably both daos and crt: crt provides info like the timeout for crt ping RPCs, while daos provides info like the expected queuing and handling times for different RPCs. Then, we may combine them together to determine the RPC timeout value. The "1/4 default" logic here does not represent what we should do eventually. The crt_req_get_timeout API, however, has a clear meaning regardless, which should remain meaningful in the future (e.g., for logging actual timeout value).

The issue I saw was that refreshing pool maps probably has become much slower with verbs, because with only a handful of engines, the chance of picking an unavailable engine to send POOL_TGT_QUERY_MAP to was not small. One such unlucky pick costed 60 s in CI testing, pushing test durations close to test timeouts. (There are a couple of additional optimizations we could do (similar to "known map versions") as proposed in the original map_refresh patch commit message. We would need to track ranks in addition to known map versions though.)

Reply:
> The crt_req_get_timeout API, however, has a clear meaning regardless, which should remain meaningful in the future (e.g., for logging actual timeout value).
The issue I saw was that refreshing pool maps probably has become much slower with verbs, because with only a handful of engines, the chance of picking an unavailable engine to send POOL_TGT_QUERY_MAP to was not small.

ok, fair enough.  Thanks for explanation.  Hmm, do you mean other server might be delay to get the updated pool map, so why not send to PS leader in this case. I do not see how shrink the timeout can help? or I miss sth?

Reply:
That is not what I meant. Suppose rank 7 becomes unavailable, a new pool map version is generated, a client sends an RPC to rank 7, the RPC times out, the client randomly picks rank 7 (unlucky) to query pool map. With verbs, this query will take a timeout, instead of failing immediately. So shrinking the query timeout speeds up this unlucky case.